spec/models/person_spec.rb

```diff
@@ -85,30 +85,37 @@ def create_role(**attrs)
     end
 
     it "with multiple membership roles returns the sum of role.membership_years" do
-      create_role(created_at: created_at, delete_on: created_at + 365.days)
+      create_role(created_at: created_at, delete_on: created_at + 1.years)
       create_role(created_at: created_at + 2.years, delete_on: created_at + 2.years + 365.days)
       expect(person_with_membership_years.membership_years).to eq 2
     end
 
-    it "rounds down to full years" do
+    it "calculates membership years correctly for the next 20 years" do
       role = create_role(delete_on: created_at + 363.days)
       expect(person_with_membership_years.membership_years).to eq 0
 
-      role.update(delete_on: created_at + 364.days)
-      expect(person_with_membership_years.membership_years).to eq 1
-
-      role.update(delete_on: created_at + 728.days)
-      expect(person_with_membership_years.membership_years).to eq 1
+      (1..20).each do |x|
+        [
+          {years_offset: x.years - 2.days, expected_years: x - 1},
+          {years_offset: x.years - 1.days, expected_years: x},
+          {years_offset: x.years, expected_years: x}
+        ].each do |test_case|
+          role.update(delete_on: role.created_at + test_case[:years_offset])
+          expect(person_with_membership_years.membership_years).to eq(test_case[:expected_years])
+        end
+      end
+    end
 
-      role.update(delete_on: created_at + 729.days)
-      expect(person_with_membership_years.membership_years).to eq 2
+    it "calculates membership years from roles starting and ending in overlaping years" do
+      role = create_role(delete_on: created_at + 200.days)
+      role.update!(created_at: created_at + 100.days)
+      expect(person_with_membership_years.membership_years).to eq 0
 
-      role.update(delete_on: created_at + 20.years)
-      expect(person_with_membership_years.membership_years).to eq 20
+      role.update(delete_on: created_at + 565.days)
+      expect(person_with_membership_years.membership_years).to eq(1)
 
-      # currently failing
-      # role.update(delete_on: created_at + 20.years - 2.days)
-      # expect(person_with_membership_years.membership_years).to eq 19
+      role.update(delete_on: created_at + 930.days)
+      expect(person_with_membership_years.membership_years).to eq(2)
```

Comment:
Wahrscheinlich wären all die Tests verständlicher, wenn mit expliziten Daten anstatt mit Tagesdifferenzen gearbeitet wird. Dabei sollten unbedingt auch diverse Start- und Enddaten verwendet werden, nicht nur der 1.1. und der 31.12. (aber auch). Ausserdem fehlen Tests, wenn ein beliebiger Stichtag mitgegeben wird. Auch zwei-drei Tests mit mehreren Rollen, welche sich an beliebigen Tagen "ablösen".

Reply:
Die Tests, wenn ein beliebiger Stichtag mitgegeben wird, fehlen weiterhin. Dort failt auch der Position Generator.

Ergänze auch noch zwei Tests mit den expliziten Gültigkeitsdaten 1.1.2020 - 31.12.2020 sowie 1.1.2020 - 1.1.2021, damit dieser Unterschied wirklich klar ist. Und dann je einmal mit dem `deleted_at` und einmal mit dem Stichtag Parameter (wobei dann das deleted_at natürlich weiter in der Zukunft liegen muss).